enterprise/cmd/repo-updater/authz/perms_scheduler.go

```diff
@@ -0,0 +1,253 @@
+ORDER BY updated_at ASC
+LIMIT %s
+`, limit)
+
+	results, err := s.scanIDsWithTime(ctx, q)
+	if err != nil {
+		return nil, err
+	}
+
+	repos := make([]ScheduledRepo, len(results))
+	for i := range results {
+		repos[i] = ScheduledRepo{
+			Priority:      PriorityLow,
+			RepoID:        api.RepoID(results[i].id),
+			LastUpdatedAt: results[i].time,
+		}
+	}
+	return repos, nil
+}
+
+// Schedule contains information for scheduling users and repositories.
+type Schedule struct {
+	Users []ScheduledUser
+	Repos []ScheduledRepo
+}
+
+// ScheduledRepo contains information for scheduling a user.
+type ScheduledUser struct {
+	Priority
+	UserID        int32
+	LastUpdatedAt time.Time
+}
+
+// ScheduledRepo contains for scheduling a repository.
+type ScheduledRepo struct {
+	Priority
+	api.RepoID
+	LastUpdatedAt time.Time
+}
+
+// schedule computes schedule four lists in the following order:
+//   1. Users with no permissions, because they can't do anything meaningful (e.g. not able to search).
+//   2. Private repositories with no permissions, because those can't be viewed by anyone except site admins.
+//   3. Rolling updating user permissions over time from oldest ones.
+//   4. Rolling updating repository permissions over time from oldest ones.
+func (s *PermsScheduler) schedule(ctx context.Context) (*Schedule, error) {
+	schedule := new(Schedule)
+
+	users, err := s.scheduleUsersWithNoPerms(ctx)
+	if err != nil {
+		return nil, errors.Wrap(err, "schedule users with no permissions")
+	}
+	schedule.Users = append(schedule.Users, users...)
+
+	repos, err := s.scheduleReposWithNoPerms(ctx)
+	if err != nil {
+		return nil, errors.Wrap(err, "schedule repositories with no permissions")
+	}
+	schedule.Repos = append(schedule.Repos, repos...)
+
+	// TODO(jchen): Predict a threshold based on total repos and users that make sense to
+	// finish syncing before next scans, so we don't waste database bandwidth.
+	// Formula (in worse case scenario, at the pace of 1 req/s):
+	//   initial threshold  = schdule interval in seconds
+	//	 consumed by users = <initial threshold> / (<total repos> / <page size>)
+	//   consumed by repos = (<initial threshold> - <consumed by users>) / (<total users> / <page size>)
+	// Hard coded both to 100 for now.
+	const threshold = 100
+
+	users, err = s.scheduleUsersWithOldestPerms(ctx, threshold)
+	if err != nil {
+		return nil, errors.Wrap(err, "load users with oldest permissions")
+	}
+	schedule.Users = append(schedule.Users, users...)
+
+	repos, err = s.scheduleReposWithOldestPerms(ctx, threshold)
+	if err != nil {
+		return nil, errors.Wrap(err, "scan repositories with oldest permissions")
+	}
+	schedule.Repos = append(schedule.Repos, repos...)
+
+	return schedule, nil
+}
+
+// Sync kicks off the permissions syncing process, this method is blocking and
+// should be called as a goroutine.
+func Sync(ctx context.Context, scheduler *PermsScheduler, syncer *PermsSyncer) {
+	go syncer.Run(ctx)
```

Comment:
when this function returns, should `syncer.Run` stop? If so maybe do something like
```suggestion
	ctx, cancel := context.WithCancel(ctx)
	defer cancel()

	go syncer.Run(ctx)
```

Reply:
Thanks, will update.